src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseCodeReviewFeedbackService.java

```diff
@@ -0,0 +1,227 @@
+                        feedback.setHasLongFeedbackText(false);
+                        feedback.setType(FeedbackType.AUTOMATIC);
+                        feedback.setCredits(0.0);
+                        return feedback;
+                    }).toList();
+
+            automaticResult.setSuccessful(true);
+            automaticResult.setCompletionDate(ZonedDateTime.now());
+
+            this.resultService.storeFeedbackInResult(automaticResult, feedbacks, true);
+
+            this.programmingMessagingService.notifyUserAboutNewResult(automaticResult, participation);
+        }
+        catch (Exception e) {
+            log.error("Could not generate feedback", e);
+            automaticResult.setSuccessful(false);
+            automaticResult.setCompletionDate(ZonedDateTime.now());
+            this.resultRepository.save(automaticResult);
+            this.programmingMessagingService.notifyUserAboutNewResult(automaticResult, participation);
+        }
+        finally {
+            programmingExerciseParticipationService.unlockStudentRepositoryAndParticipation(participation);
+            participation.setIndividualDueDate(null);
+            this.programmingExerciseStudentParticipationRepository.save(participation);
+        }
+    }
+
+    /**
+     * Sets an individual due date for a participation, locks the repository,
+     * and invalidates previous results to prepare for new feedback.
+     *
+     * @param participation       the programming exercise student participation.
+     * @param programmingExercise the associated programming exercise.
+     */
+    private ProgrammingExerciseStudentParticipation setIndividualDueDateAndLockRepository(ProgrammingExerciseStudentParticipation participation,
+            ProgrammingExercise programmingExercise, boolean invalidatePreviousResults) {
+        var currentDate = now();
+
+        // The participations due date is a flag showing that a feedback request is sent
+        participation.setIndividualDueDate(currentDate);
+
+        participation = programmingExerciseStudentParticipationRepository.save(participation);
+        // Circumvent lazy loading after save
+        participation.setParticipant(participation.getParticipant());
+        programmingExerciseParticipationService.lockStudentRepositoryAndParticipation(programmingExercise, participation);
+
+        if (invalidatePreviousResults) {
+            var participationResults = participation.getResults();
+            participationResults.forEach(participationResult -> participationResult.setRated(false));
+            this.resultRepository.saveAll(participationResults);
+        }
+
+        return participation;
+    }
+
+    private void checkRateLimitOrThrow(ProgrammingExerciseStudentParticipation participation) {
+
+        var athenaResults = participation.getResults().stream().filter(result -> result.getAssessmentType() == AssessmentType.AUTOMATIC_ATHENA).toList();
```

Comment:
According to the [server coding guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server.html#naming-convention), this should also have an explicit type. So should have the following integers.